proxy/src/serve_sockets.rs

```diff
@@ -0,0 +1,523 @@
+            return Ok(None);
+        }
+
+        let plain = self
+            .decryptor
+            .decrypt_next(&src[Self::SIZE_OVERHEAD..])
+            .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "Decryption failed"))?;
+        src.advance(total_frame_size);
+        Ok(Some(plain))
+    }
+}
+
+struct EncBuffer<'a> {
+    buf: &'a mut BytesMut,
+}
+
+impl<'a> EncBuffer<'a> {
+    fn new(buffer: &'a mut BytesMut, content_len: usize) -> Self {
+        buffer.reserve(content_len + Self::SIZE_OVERHEAD);
+        buffer.extend_from_slice(&u32::to_le_bytes(content_len as u32));
+        Self { buf: buffer }
+    }
+
+    /// Reserved for size of msg
+    const SIZE_OVERHEAD: usize = 4;
+}
+
+impl<'a> Buffer for EncBuffer<'a> {
+    // This should only be called to append the tag to the buffer
+    fn extend_from_slice(&mut self, other: &[u8]) -> aead::Result<()> {
+        self.buf.extend_from_slice(other);
+        Ok(())
+    }
+
+    // This should only be called when decrypting
+    fn truncate(&mut self, len: usize) {
+        self.buf.truncate(len)
+    }
+}
+
+impl<'a> AsRef<[u8]> for EncBuffer<'a> {
+    fn as_ref(&self) -> &[u8] {
+        &self.buf[Self::SIZE_OVERHEAD..]
+    }
+}
+
+impl<'a> AsMut<[u8]> for EncBuffer<'a> {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.buf[Self::SIZE_OVERHEAD..]
+    }
+}
+
+impl<S: AsyncRead + AsyncWrite + Unpin> EncryptedSocket<S> {
+    /// Creates a new cipher stream with a 32 byte key and a 16 + 4 byte Nonce
+    async fn new(mut inner: S, key: &GenericArray<u8, U32>) -> io::Result<Self> {
+        let aead = XChaCha20Poly1305::new(key);
+
+        // Encryption
+        let mut enc_nonce = GenericArray::default();
+        OsRng.fill_bytes(&mut enc_nonce);
+        let encryptor = EncryptorLE31::from_aead(aead.clone(), &enc_nonce);
+        inner.write_all(&enc_nonce).await?;
+
+        // Decryption
+        let mut dec_nonce = GenericArray::default();
+        inner.read_exact(dec_nonce.as_mut_slice()).await?;
+        let decryptor = DecryptorLE31::from_aead(aead, &dec_nonce);
+
+        let (r, w) = tokio::io::split(inner);
+        let read = FramedRead::new(r, DecryptorCodec { decryptor });
+        let read = read.into_async_read().compat();
+        let write = FramedWrite::new(w, EncryptorCodec { encryptor });
+
+        Ok(Self { read, write })
+    }
+}
+
+impl<S: AsyncRead + AsyncWrite + Unpin> AsyncRead for EncryptedSocket<S> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<io::Result<()>> {
+        Pin::new(&mut self.read).poll_read(cx, buf)
+        // self.read.poll_next_unpin(cx).map(|item| match item {
+        //     Some(Ok(plain)) => {
+        //         buf.put_slice(&plain);
+        //         Ok(())
+        //     },
+        //     Some(Err(e)) => Err(e),
+        //     None => Ok(())
+        // })
```

Comment:
Please remove dead code

Reply:
Yeah these are relics from the suffering I went through to make the encryption work.